Do not reallocate whole input string when matching next token

mvorisek opened the pull request:
Purely a performance refactoring.

The measured performance improvement is about -90% reduced runtime for the included test. With larger test/SQL, the speedup is even more dramatic as the original complexity was O(N^2).

Originally, the whole input string was reallocated after each token in:
- https://github.com/doctrine/sql-formatter/blob/1.4.0/src/Tokenizer.php#L790

such big temporary string was upper cased for each token matching in:
- https://github.com/doctrine/sql-formatter/blob/1.4.0/src/Tokenizer.php#L890

~and regex patterns were constructed for each token matching freshly in:~ (extracted into another PR)
- https://github.com/doctrine/sql-formatter/blob/1.4.0/src/Tokenizer.php#L874

All such operations should be avoided for linear scalability and this PR addresses that.

mvorisek committed:
Refactor all "strtoupper($stringSlow)" code

mvorisek commented:
PR is finished and the performance improvement (of the "long concat" tests with new 20k elements limit) is the following:

before:
```
21.370 secs Doctrine\SqlFormatter\Tests\SqlFormatterTest::testFormatLongConcat
18.474 secs Doctrine\SqlFormatter\Tests\TokenizerTest::testTokenizeLongConcat
```

after
```
5.710 secs Doctrine\SqlFormatter\Tests\SqlFormatterTest::testFormatLongConcat
1.665 secs Doctrine\SqlFormatter\Tests\TokenizerTest::testTokenizeLongConcat
```

So the tokenization process is newly about 10x faster and with increased query length the speedup is even bigger.

This PR is faster also on the `tests/performance.php` script even it uses very small queries:

before:
```
$ tests\performance.php                        
    <p>Formatted 54 queries</p>
    <p>Average query length of 284.37037 characters</p>
    <p>
        Took 0.02518 seconds total,
        0.00047 seconds per query,
        0.00164 seconds per 1000 characters
    </p>
    <p>Used 0 bytes of memory</p>

(2nd run)
$ tests\performance.php
    <p>Formatted 54 queries</p>
    <p>Average query length of 284.37037 characters</p>
    <p>
        Took 0.02403 seconds total,
        0.00044 seconds per query,
        0.00156 seconds per 1000 characters
    </p>
    <p>Used 0 bytes of memory</p>
```

after:
```
$ tests\performance.php
    <p>Formatted 54 queries</p>
    <p>Average query length of 284.37037 characters</p>
    <p>
        Took 0.01587 seconds total,
        0.00029 seconds per query,
        0.00103 seconds per 1000 characters
    </p>
    <p>Used 0 bytes of memory</p>

(2nd run)
$ tests\performance.php
    <p>Formatted 54 queries</p>
    <p>Average query length of 284.37037 characters</p>
    <p>
        Took 0.01543 seconds total,
        0.00029 seconds per query,
        0.00101 seconds per 1000 characters
    </p>
    <p>Used 0 bytes of memory</p>
```

goetas commented on src/Tokenizer.php:
```diff
@@ -726,74 +727,108 @@ final class Tokenizer
-        $token = null;
-
-        $currentLength = strlen($string);
+        $upper  = strtoupper($string);
```
why all uppercase?

mvorisek commented:
@greg0ire can I please have your review here?

goetas commented on src/Tokenizer.php:
```diff
@@ -726,57 +727,108 @@ final class Tokenizer
+    }
+
+    /** @param list<string> $values */
+    private function makeRegexFromList(array $values, bool $sorted = false): string
```
i really think that this is an overkill. i see here an attempt to create an abstraction ... and suddenly a 4 lines closure became a 60 lines private method

mvorisek replied:
This gives 13% speedup in average and much more in edge usecases. I stand strongly by this regex generation and it is a technique commonly done by libs like repan - https://github.com/zherczeg/repan/blob/25d6e3aabf/tests/opt/test_merge_alternatives_expected.txt.

goetas commented on src/Token.php:
```diff
@@ -37,6 +40,18 @@ public function value(): string
         return $this->value;
     }
 
+    public function valueUpper(): string
```
i'm really not sure this method should be here... this is just an implementation detail for your tokenizer implementation. i think that a `strtoupper` was fine

mvorisek replied:
This has currently small (1.5%) speedup, I personally like this improvement as the code is then simpler and faster. As the `Token` is more or less internal class, I do not see any problem with this well defined and handly method.

SenseException replied:
Do we even need to invoke `strtoupper($token->value())` all the time instead of using a variable? It would make `valueUpper()` obsolete.

mvorisek replied:
We need - as we need upper case prev/next token too.

SenseException replied:
The amount of the previous token's value is manageable. Less method and functions invocations would also be in the interest of your previous reasoning.

mvorisek replied:
I have reverted the `Token::valueUpper()` improvement for now as it is unrelated to the main optimization done by this PR.

goetas reviewed (changes requested):
I can understand the desire to optimize performance, but I also suggest to keep in mind readability and maintainability of the new code.

Do you have a specific use case that you are solving with this lib that you need to squeeze every CPU cycle?

goetas commented on src/Tokenizer.php:
```diff
@@ -726,57 +727,108 @@ final class Tokenizer
         $tokens = [];
 
-        $token = null;
+        $upper  = strtoupper($string);
```
given that this optimizations in this method are begin kept, do you have a benchmark to understand the impact of all the other optimizations you are suggesting?

mvorisek replied:
Previously the uppercased matching was already done... and done very poorly as the string was upper cased again and again. The overall speedup of this PR is more than -90% runtime (10x faster) - see https://github.com/doctrine/sql-formatter/actions/runs/9366932237/job/25785513399?pr=119#step:7:25 now and before. This is discussed in https://github.com/doctrine/sql-formatter/pull/119#issuecomment-2143747858 and the PR description.

greg0ire commented:
> Do you have a specific use case that you are solving with this lib that you need to squeeze every CPU cycle?

I'm also curious about this.

mvorisek commented:
> > Do you have a specific use case that you are solving with this lib that you need to squeeze every CPU cycle?
> 
> I'm also curious about this.

We started using this package in https://github.com/atk4/data/pull/1216 where we use this package to format log and debug queries. When I run all tests with this package, I noticed quite massive slowdown so I started investigating...

There were these issues:

1. constructor slowness - the class/grammar is final, we can use static property safely to build the regexes only once
2. reallocation - the solution is to use offset and `\G` instead of `^` for matching offset start
3. non-optimal regexes from lists generated - see https://3v4l.org/9jitR/rfc#vgit.master_jit for real improvent times

mvorisek commented:
I hope the massive overall speedup and linear scalability is well enough to justify this PR. Can this PR be reviewed?

Performance is compared in https://github.com/doctrine/sql-formatter/pull/119#issuecomment-2143747858.

mvorisek commented:
Please, can this PR be reviewed.

mvorisek commented:
Guys, can we please move into integrating this PR?

greg0ire commented:
Way too many commits, way too little explanations… let's take an example among many:

> Change "^" to "\G" in regexes to match offset start 

Are we supposed to know why the hell one would want to do that?

mvorisek commented:
> Way too many commits, way too little explanations… let's take an example among many:
> 
> > Change "^" to "\G" in regexes to match offset start
> 
> Are we supposed to know why the hell one would want to do that?

`\G` is documented in https://www.php.net/manual/en/regexp.reference.escape.php - I have now added also a commit message.

Anything else?

greg0ire commented on src/Tokenizer.php:
```diff
@@ -822,20 +822,23 @@ public function tokenize(string $string): Cursor
-    private function createNextToken(string $string, Token|null $previous = null): Token
+    private function createNextToken(string $string, string $upper, int $offset, Token|null $previous = null): Token
     {
+        $stringSlow = substr($string, $offset);
```
`stringSlow`? Why did you pick that name?

mvorisek replied:
Purely for "inner PR" refactoring, it is much more readable to have several "inner PR" refactoring commits doiing one change at a time.

greg0ire commented:
> Refactor first/second char matching 

How did you refactor it and why?

greg0ire commented:
Please split this; there seems to be changes that could be indenpendent, for instance "Remove historical comments" can be split out.

mvorisek commented:
> > Refactor first/second char matching
> 
> How did you refactor it and why?

Previously the `$str[0]` was refering to the first char, but this must be avoided, as this required many reallocations `$str = substr($str, $offset)`. Newly, we do not reallocate the string but instead of use `$offset` everywhere, this the first char is `$str[$offset]` newly, and for easier readability, I changed `$str[0]` to `$str[$offset]`. Earlier refactoring commits make sense for the next commits.

mvorisek commented:
> Please split this; there seems to be changes that could be indenpendent, for instance "Remove historical comments" can be split out.

commit removed

greg0ire commented:
Are there ways you could split this in several smaller PRs?

mvorisek commented:
> Are there ways you could split this in several smaller PRs?

Yes - done now, here we focus on "no reallocation", ie. using `$offset` strictly.

greg0ire commented:
Is 3c2a6db79e2001c2ec89b1a81a3296252cf0e97d about avoiding allocations? It seems to allocate more variables.

mvorisek commented:
> Is [3c2a6db](https://github.com/doctrine/sql-formatter/commit/3c2a6db79e2001c2ec89b1a81a3296252cf0e97d) about avoiding allocations? It seems to allocate more variables.

Such short allocations are fast and more readable over `$str[$offset]` and `$str[$offset + 1]`. Especially, single char string is always interned thus no real allocation is done. This is thanks to internal php optimization.

However I tested it - offset is slightly faster over `substr($str, $offset, 1)`. I belive this is because specific opcode is used over regular function call/overhead.

greg0ire commented:
Since it is not related (and even the opposite of what this PR is about, reducing allocations), then let's move the 2 commits in a separate PR please.

mvorisek commented:
Like really, do I have to use separate PR for using separate variable when this PR is refactoring this topic and the performance was tested?

greg0ire commented:
Do I have to deal with PRs with 7 commits? 7, really? What the hell?

greg0ire reviewed (changes requested):
This PR had better be strictly about removing allocations, not adding some.

mvorisek committed:
Refactor Tokenizer::createNextToken() to accept string and offset

mvorisek committed:
Change "^" to "\G" in regexes to match offset start

`^` matches "string start", `\G` is the same but matches start given by the 5th `preg_match` `$offset` argument.

mvorisek committed:
Refactor Tokenizer::getNextQuotedString() to accept string and offset

mvorisek committed:
Refactor all remaining "$stringSlow" code

mvorisek commented:
done

greg0ire commented:
Thanks @mvorisek !
